massa-protocol-worker/src/connectivity.rs

```diff
@@ -0,0 +1,242 @@
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::{num::NonZeroUsize, sync::Arc};
+use std::{thread::JoinHandle, time::Duration};
+use tracing::warn;
+
+use crate::handlers::peer_handler::PeerManagementHandler;
+use crate::{handlers::peer_handler::models::SharedPeerDB, worker::ProtocolChannels};
+use crate::{
+    handlers::{
+        block_handler::{cache::BlockCache, BlockHandler},
+        endorsement_handler::{cache::EndorsementCache, EndorsementHandler},
+        operation_handler::{cache::OperationCache, OperationHandler},
+        peer_handler::models::PeerMessageTuple,
+    },
+    wrap_network::NetworkController,
+};
+
+pub enum ConnectivityCommand {
+    Stop,
+    GetStats {
+        #[allow(clippy::type_complexity)]
+        responder: Sender<(
+            NetworkStats,
+            HashMap<PeerId, (SocketAddr, PeerConnectionType)>,
+        )>,
+    },
+}
+
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn start_connectivity_thread(
+    config: ProtocolConfig,
+    mut network_controller: Box<dyn NetworkController>,
+    consensus_controller: Box<dyn ConsensusController>,
+    pool_controller: Box<dyn PoolController>,
+    channel_blocks: (Sender<PeerMessageTuple>, Receiver<PeerMessageTuple>),
+    channel_endorsements: (Sender<PeerMessageTuple>, Receiver<PeerMessageTuple>),
+    channel_operations: (Sender<PeerMessageTuple>, Receiver<PeerMessageTuple>),
+    channel_peers: (Sender<PeerMessageTuple>, Receiver<PeerMessageTuple>),
+    bootstrap_peers: Option<BootstrapPeers>,
+    peer_db: SharedPeerDB,
+    storage: Storage,
+    protocol_channels: ProtocolChannels,
+) -> Result<(Sender<ConnectivityCommand>, JoinHandle<()>), ProtocolError> {
+    let initial_peers = if let Some(bootstrap_peers) = bootstrap_peers {
+        bootstrap_peers.0.into_iter().collect()
+    } else {
+        serde_json::from_str::<HashMap<PeerId, HashMap<SocketAddr, TransportType>>>(
+            &std::fs::read_to_string(&config.initial_peers)?,
+        )?
+    };
+
+    let handle = std::thread::Builder::new()
+    .name("protocol-connectivity".to_string())
+    .spawn({
+        let sender_endorsements_propagation_ext = protocol_channels.endorsement_handler_propagation.0.clone();
+        let sender_blocks_retrieval_ext = protocol_channels.block_handler_retrieval.0.clone();
+        let sender_blocks_propagation_ext = protocol_channels.block_handler_propagation.0.clone();
+        let sender_operations_propagation_ext = protocol_channels.operation_handler_propagation.0.clone();
+        move || {
+            for (addr, transport) in &config.listeners {
+                network_controller
+                    .start_listener(*transport, *addr)
+                    .unwrap_or_else(|_| panic!(
+                        "Failed to start listener {:?} of transport {:?} in protocol",
+                        addr, transport
+                    ));
+            }
+            std::thread::sleep(Duration::from_millis(100));
```

Comment:
add // Hack to explain why we sleep here

Reply:
To prevent removing by someone

Reply:
Done